zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1147 @@
+{
+	const struct sc_can_cfg *config = dev->config;
+
+	*rate = config->clock_frequency;
+
+	return 0;
+}
+
+/*
+ * This API specification is here:
+ *   Get the maximum standard (11-bit) CAN ID filters if false, or extended (29-bit)
+ *   CAN ID filters if true
+ * Acceptance filter of SC CAN Controller supports both standard and extended, so
+ * always returns the same value.
+ */
+static int sc_can_get_max_filters(const struct device *dev, bool ide)
+{
+	const struct sc_can_cfg *config = dev->config;
+
+	ARG_UNUSED(ide);
+
+	return config->max_filter;
+}
+
+static int sc_can_get_max_bitrate(const struct device *dev, uint32_t *max_bitrate)
+{
+	const struct sc_can_cfg *config = dev->config;
+
+	*max_bitrate = config->max_bitrate;
+
+	return 0;
+}
+
+static const struct can_driver_api sc_can_driver_api = {
+	.get_capabilities = sc_can_get_capabilities,
+	.start = sc_can_start,
+	.stop = sc_can_stop,
+	.set_mode = sc_can_set_mode,
+	.set_timing = sc_can_set_timing,
+	.send = sc_can_send,
+	.add_rx_filter = sc_can_add_rx_filter,
+	.remove_rx_filter = sc_can_remove_rx_filter,
+	.get_state = sc_can_get_state,
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+	.recover = sc_can_recover,
+#endif
+	.set_state_change_callback = sc_can_set_state_change_callback,
+	.get_core_clock = sc_can_get_core_clock,
+	.get_max_filters = sc_can_get_max_filters,
+	.get_max_bitrate = sc_can_get_max_bitrate,
+	/*
+	 * ISO 11898-1:
+	 *   Nominal length of time segments (in absence of synchronization)
+	 *   - Sync_Seg shall be one time quantum long.
+	 *   - The information processing time shall be less than or equal to two time quanta long.
+	 *   - Prop_Seg shall be programmable to be 1, 2, 3, . . ., 8 or more time quanta long.
+	 *     It shall be programmed to compensate the delay times of the actual network,
+	 *     rounded up to the nearest integer time quantum.
+	 *   - Phase_Seg1 shall be programmable to be 1, 2, 3, . . ., 8 or more time quanta long.
+	 *   - Phase_Seg2 shall be programmed to the maximum of Phase_Seg1 and the information processing
+	 *     time.
+	 *   - SJW shall be programmable between 1 and the minimum of Phase_Seg1 and 4.
+	 *
+	 *   In a CAN implementation, Prop_Seg and Phase_Seg1 do not need to be programmable separately;
+	 *   it is sufficient to program the sum of Prop_Seg and Phase_Seg1.
+	 *   The total number of time quanta in a bit time shall be programmable at least from 8 to 25.
+	 * In this driver:
+	 *  - SJW
+	 *    set to 1 - 4.
+	 *  - prop_seg / phase_seg1
+	 *    SC CAN IP core does not separate the Prop_Seg and Phase_Seg1,so prop_seg is set to 0,
+	 *    and phase_seg1 is set to 2 - 16
+	 *    (Sum of minimum/maximum value of Prop_Seg and Phase_Seg1)
+	 *  - phase_seg2
+	 *    set to 1 - 8.
+	 *  - prescaler
+	 *    SC CAN IP core have 16 bit register value for set the prescaler. so set to 0x1000.
+	 */
+	.timing_min = { .sjw = 0x1,
+			.prop_seg = 0x00,
```

Comment:
ISO says "Prop_Seg shall be programmable to be 1, 2,...".  You can't use `0` for it.

Reply:
Yes, you are right. 
I misunderstood specification of prop_seq and phase_seg1.
Each setting should be set between 1 and 8.